readme.md
# npm-downloads-counter

This is a [CLI](https://en.wikipedia.org/wiki/Command-line_interface) tool (and an [ES module](https://nodejs.org/api/esm.html)) to output an overview of the (last 30 days) download activity of all the packages maintained (or published) by a certain [NPM](https://www.npmjs.com/) username. I've also added support for outputting the score statistics of each package.

The tool can even push notifications (containing this output) to your Slack channel whenever there are new statistics available when the tool is ran. This could be used together with e.g. [cron](https://en.wikipedia.org/wiki/Cron) or [systemd timers](https://opensource.com/article/20/7/systemd-timers) for daily updates.

I created this mainly because I wanted an easy way to monitor the popularity of the [packages I've published](https://www.npmjs.com/~joakimch).

## Funding

If you find this useful then please consider helping me out (I'm jobless and sick). For more information visit my [GitHub profile](https://github.com/JoakimCh).

## Cross-platform support

I've tested it successfully on Linux (my main system), Windows and macOS.

##### Windows usage notes

For the unicode output of this program to display correctly you'll need to use the new [Windows Terminal](https://www.microsoft.com/en-us/p/windows-terminal/9n0dx20hk701).

## Installing

Install it globally like this:
```bash
npm install -g npm-downloads-counter
```
Optionally install [chalk](https://www.npmjs.com/package/chalk) to enable colors in CLI:
```bash
npm i -g chalk
```
Optionally install [terminal-link](https://www.npmjs.com/package/terminal-link) to enable links in CLI:
```bash
npm i -g terminal-link
```

## How to use

To list all packages published by a user (e.g. to list packages published by yourself):
```bash
npm-downloads-counter npm_username
```

To list all packages maintained by a user (a package can have several maintainers though):
```bash
npm-downloads-counter -m npm_username
```

To list (only) score statistics:
```bash
npm-downloads-counter --score npm_username
```

You can also run it without installing it using [npx](https://docs.npmjs.com/cli/v7/commands/npx), but then it can't have colors or links in the CLI:
```bash
npx npm-downloads-counter joakimch
```

## Example output (download stats)
```
Download count of packages published by joakimch the last 30 days:
|⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣾⣦⣀⣿⢀⠀⠀⠀⢀⠀| npm-downloads-counter 60 0-22 
|⢀⠀⠀⠀⣀⣠⢀⢀⠀⠀⠀⠀⣿⢀⢀⠀⠀⠀⠀⢀⠀⠀⠀⠀⣿⢀⠀⠀⠀⢀| ▲1 pluggable-prng 59 0-20 
|⠀⢀⠀⢀⢀⢀⢀⣾⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⠀⣿⢀⠀⢀⢀⢀⠀⠀⠀⠀⢀| ▼1 bit-consumer 57 0-22 
|⠀⠀⠀⢀⢀⠀⠀⣿⠀⠀⠀⠀⠀⠀⠀⢀⠀⠀⠀⣿⢀⠀⠀⢀⢀⠀⢀⠀⠀⠀| base-endecoder 45 0-20 
|⠀⠀⠀⠀⢀⢀⠀⣾⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣿⠀⠀⠀⢀⠀⠀⠀⠀⠀⢀| ▲1 bit-manipulation 44 0-21 
|⢀⠀⠀⠀⢀⠀⠀⠀⠀⠀⠀⢀⣿⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⣿⠀⢀⠀⢀⠀| ▼1 platform-checker 44 0-20 
|⠀⠀⠀⢀⢀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⠀⠀⠀⠀⢀⢀⠀⠀⣿⠀⠀⠀⠀⠀| whichever-data 30 0-24 

Summary of packages climbing the list:
pluggable-prng is up: 1
bit-manipulation is up: 1
```

## Output explained

I've designed the output from this tool to be as minimal as it can get to easily fit in most terminal windows. Hence I feel the need to explain what you're looking at.

Each line represents a different package.

It begins with a bar graph (cleverly using unicode braille characters) to represent the "download pressure" of the last 30 days. Where a space represents a download count that day equal to the least amount of daily downloads of this period. Whereas a ⣿ symbol represents the maximum amount (but is not used when the maximum is less than 8 downloads).

The graph is followed by the package name, which is then followed by a number representing the total of downloads this period and some min-max numbers representing the minimum and maximum amount of daily downloads this period.

The lines are sorted so that the packages on top are the ones with the highest amount of downloads this period.

If the package has changed its position in the list compared to when you last ran the tool then it will be indicated by a ▲ or ▼ before the package name, together with a number representing how many lines it jumped from.


## Example output (score stats)
```
Score list of packages published by joakimch sorted by final:
bit-consumer {
  final: 0.2816297020884584,
  quality: 0.5214210083588049,
  popularity: 0.024733028932926553,
  maintenance: 0.33299096986940757,
  searchScore: 0.00000006146252
}
whichever-data {
  final: 0.2810230954009417,
  quality: 0.5214210083588049,
  popularity: 0.022999866968593137,
  maintenance: 0.33299096986940757,
  searchScore: 0.000000058837163
}
[and so on...]
```

## Custom configuration

This tool supports some configuration, mainly of the output format of the statistics. Below is the code that loads the config file, you can use this as an overview of what can be written in the `config.json` file.

Tip, use `--paths` to output the `configDir` path on your platform; this is where to place the `config.json` file.
```js
const options = { // merge default with custom config
  // noColors: true, // no colors in CLI even if chalk is installed
  // noLinks: true, // no links in CLI even if terminal-link is installed
  // separateLines: true, // graph is on a separate line
  // thinBars: true, // use thin bars in graph
  displayDate: true, // date of when download-statistics was updated
  logToTerminal: true, // output the result to terminal
  climbersSummary: true, // summary of packages climbing the list
  comparePrevious: true, // show position changes of packages
  ...loadConfig(configDir+'config.json') // merge (overwriting options above)
}
if (options.slack) {
  options.slack = { // Slack only options
    // token: 'xoxb-XXXX-XXXX', // the much needed Web API token
    // channel: 'npm-downloads-counter', // which channel to post to, name or id
    // username: 'npm-downloads-counter', // custom username for message
    // icon_emoji: ':bar_chart:', // custom icon for message
    // noScores: true, // do not push notifications about scores
    // onlyUsername: 'username', // only push notifications when checking this username
    thinBars: true, // use thin bars in graph
    separateLines: true, // graph is on a separate line
    ...options.slack, // merge (overwriting options above)
    noColors: true, // not compatible with Slack
    noLinks: true // not compatible with Slack
  }
}
```

## Slack notifications

### How to setup a new Slack app for notifications

Go to https://api.slack.com/apps/ and press "Create New App", then "From scratch".

Enter the page for your app and under "Add features and functionality" press "Permissions" then scroll down to "Scopes" and add these "Bot Token Scopes" by pressing "Add an OAuth Scope":
* `chat:write.public` To send messages to channels it isn't a member of.
* `chat:write.customize` To send messages with a customized username and avatar.

Then you can scroll up to "OAuth Tokens for Your Workspace" and press "Install to Workspace". This results in a "Bot User OAuth Token" which is the token you need to use. It should look similar to this:
`xoxb-1234567891230-9876543219870-theAlPhaBetRoCks1234Abcz`

Then in your `config.json` put something like this to use it:
```json
{
  "slack": {
    "token": "xoxb-1234567891230-9876543219870-theAlPhaBetRoCks1234Abcz",
    "channel": "the-channel-to-post-in"
  }
}
```

I find it useful to also add:
```json
"onlyUsername": "your_NPM_username"
```
So that you can check the download and score statistics of other NPM-usernames without having them pushed to your Slack channel.

## Usage as an ES module

I didn't design this to be a nice and useful library (I only care about the CLI tool), but I have exported the functions used by the CLI tool so that anyone can use them. Check the documentation below to learn about them.

## Module documentation

### Auto-generated documentation from JSDoc

## Classes

<dl>
<dt><a href="#PackageStatistics">PackageStatistics</a></dt>
<dd><p>Used to keep track of and work with the statistics collected by this library. I will not document its functions very well, instead check the code for usage examples.</p>
</dd>
</dl>

## Functions

<dl>
<dt><a href="#getPackages">getPackages(maintainerUsername, [publisherUsername])</a> ⇒ <code>Array.&lt;object&gt;</code></dt>
<dd><p>Get all packages where <code>maintainerUsername</code> is listed as one of the maintainers (anyone publishing a package is automatically listed as one). Optionally filter out any packages not published by <code>publisherUsername</code> (this can be used to return only packages published by this user).</p>
</dd>
<dt><a href="#getLastDownloads">getLastDownloads(packages, range)</a> ⇒ <code>object</code> | <code>Array.&lt;object&gt;</code></dt>
<dd><p>Get the daily download counts of a package (or packages) in the given timeframe. If doing a bulk query then limit it to &lt;=128 packages.</p>
</dd>
<dt><a href="#getStatistics">getStatistics(options)</a> ⇒ <code><a href="#PackageStatistics">PackageStatistics</a></code></dt>
<dd><p>Get either download or score statistics of all packages where <code>maintainerUsername</code> is listed as one of the maintainers (anyone publishing a package is automatically listed as one). Optionally filter out any packages not published by <code>publisherUsername</code> (this can be used to return only packages published by this user).</p>
</dd>
<dt><a href="#pushToSlack">pushToSlack(options)</a> ⇒ <code>Promise.&lt;object&gt;</code></dt>
<dd><p>Push provided text to the channel, with customizable username and icon emoji.</p>
</dd>
</dl>

<a name="getPackages"></a>

## getPackages(maintainerUsername, [publisherUsername]) ⇒ <code>Array.&lt;object&gt;</code>
Get all packages where `maintainerUsername` is listed as one of the maintainers (anyone publishing a package is automatically listed as one). Optionally filter out any packages not published by `publisherUsername` (this can be used to return only packages published by this user).

**Kind**: global function  
**Returns**: <code>Array.&lt;object&gt;</code> - An array of objects with details about the packages.  

| Param | Type | Description |
| --- | --- | --- |
| maintainerUsername | <code>string</code> | Get all packages listed with this user as one of its maintainers. |
| [publisherUsername] | <code>string</code> | Return only packages with this user listed as the publisher. |

<a name="getLastDownloads"></a>

## getLastDownloads(packages, range) ⇒ <code>object</code> \| <code>Array.&lt;object&gt;</code>
Get the daily download counts of a package (or packages) in the given timeframe. If doing a bulk query then limit it to <=128 packages.

**Kind**: global function  
**Returns**: <code>object</code> \| <code>Array.&lt;object&gt;</code> - {packageName, dailyDownloads} (in an array if doing a bulk query)  

| Param | Type | Description |
| --- | --- | --- |
| packages | <code>string</code> \| <code>Array.&lt;string&gt;</code> | The package (or packages if an array) to query. |
| range | <code>string</code> | A range/timeframe compatible with the registry API. |

<a name="getStatistics"></a>

## getStatistics(options) ⇒ [<code>PackageStatistics</code>](#PackageStatistics)
Get either download or score statistics of all packages where `maintainerUsername` is listed as one of the maintainers (anyone publishing a package is automatically listed as one). Optionally filter out any packages not published by `publisherUsername` (this can be used to return only packages published by this user).

**Kind**: global function  

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| options | <code>object</code> |  |  |
| options.maintainerUsername | <code>string</code> |  |  |
| options.publisherUsername | <code>string</code> |  |  |
| [options.range] | <code>string</code> | <code>&quot;&#x27;last-month&#x27;&quot;</code> | A range/timeframe compatible with the registry API (used for download stats). |
| [options.withoutDownloads] | <code>boolean</code> |  | To only get score stats. |

<a name="pushToSlack"></a>

## pushToSlack(options) ⇒ <code>Promise.&lt;object&gt;</code>
Push provided text to the channel, with customizable username and icon emoji.

**Kind**: global function  
**Returns**: <code>Promise.&lt;object&gt;</code> - The result from the Slack Web API.  

| Param | Type | Default |
| --- | --- | --- |
| options | <code>object</code> |  | 
| options.token | <code>string</code> |  | 
| options.channel | <code>string</code> |  | 
| options.text | <code>string</code> |  | 
| [options.username] | <code>string</code> | <code>&quot;&#x27;npm-downloads-counter&#x27;&quot;</code> | 

changelog.md
# Changelog

Changelog for the [npm-downloads-counter](https://www.npmjs.com/package/npm-downloads-counter) NPM package.

## v2.0.3 - 2021.12.06:

### Fixed

* Fixed crash when a package without stats (newly published) were discovered.


## v2.0.2 - 2021.07.26:

### Fixed

* Removed some debug output mixed with the output of --score.


## v2.0.1 - 2021.07.26:

### Changed

* Updated the readme with details on how to get colors, etc...
* The output of --paths on Windows to only use forward slashes (Windows is compatible with them yes).


## v2.0.0 - 2021.07.23:

Some new features. E.g. support for Slack notifications and the output of score statistics.

### Added

* Support for comparing the latest statistics with the previous statistics to indicate which packages are climbing or falling in the list. With a summary of which packages climbed to higher positions. (on by default)

* Support for outputting score-statistics instead of download-statistics (`--score`).

* Support for pushing notifications to your [Slack](https://slack.com/) channel of choice whenever you run the CLI and it downloaded updated statistics (e.g. to receive daily download and/or score stats).

* Support for changing some options (mostly in regards to the output) by editing a config.json file (this is also where you would configure the Slack notifications).

* Support for a smaller bar chart (optional) where each braille symbol can represent two bars (two days).

* CLI colors enabled when [chalk](https://www.npmjs.com/package/chalk) is detected¹ (can be turned off).

* CLI links enabled when [terminal-link](https://www.npmjs.com/package/terminal-link) is detected¹ (can be turned off).

*1: Detected as in installed as a dependency or globally available package.*

### Fixed

* In the downloads-graph use U+2800 (Braille Pattern Blank) instead of space to maintain a monospaced (fixed width) graph.

* A bug limiting the package list to only the most 20 popular packages.


## v1.0.0 - 2021.07.08:

Initial release.

Basic functionality for outputting a list showing the download activity of every package with the specified NPM username listed as author or maintainer.


source/npm-downloads-counter.js
/*
Documentation:
  Registry API:    https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md
  Download counts: https://github.com/npm/registry/blob/master/docs/download-counts.md
*/

import * as https from 'https'
import * as fs from 'fs'
import * as braille from './braille.js'
let Chalk; try {Chalk = (await import('chalk')).default} catch {} // optional dependency (for colors)
let terminalLink; try {terminalLink = (await import('terminal-link')).default} catch {} // optional dependency (for links in terminal)

const wideNumber = new Intl.NumberFormat('fullwide', {maximumSignificantDigits: 21})
const maxConcurrentDownloads = 10

function getJson(url) {
  return new Promise((resolve, reject) => {
    https.request(url, {}, response => {
      if (response.statusCode != 200) reject(Error('Response code: '+response.statusCode))
      let json = ''
      response.setEncoding('utf-8')
      response.on('data', chunk => json += chunk)
      response.on('end', () => {
        try {resolve(JSON.parse(json))} 
        catch (error) {reject(error)}
      })
    })
    .on('error', reject)
    .end()
  })
}

async function getPackages_(maintainerUsername, offset, limit) {
  if (limit > 250) throw Error('Max 250 results per query, reduce limit below this.')
  const queryResult = await getJson('https://registry.npmjs.com/-/v1/search?text=maintainer:'+maintainerUsername+'&size='+limit+'&from='+offset)
  if ('error' in queryResult) throw Error('API error: '+queryResult.error)
  return [queryResult.total, queryResult.objects.map(v => {return {...v.package, score: v.score, searchScore: v.searchScore}})]
}

/** Get all packages where `maintainerUsername` is listed as one of the maintainers (anyone publishing a package is automatically listed as one). Optionally filter out any packages not published by `publisherUsername` (this can be used to return only packages published by this user).
 @param {string} maintainerUsername Get all packages listed with this user as one of its maintainers.
 @param {string} [publisherUsername] Return only packages with this user listed as the publisher.
 @returns {Array.<object>} An array of objects with details about the packages.
*/
export async function getPackages(maintainerUsername, publisherUsername) {
  // https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md#get-v1search
  // they are by default sorted by popularity it seems
  const downloadPromises = []
  let limit = 250, packages = []
  const [packageCount, page1] = await getPackages_(maintainerUsername, 0, limit)
  packages.push(...page1)
  if (packageCount > limit) {
    for (let offset = limit; offset < packageCount; offset += limit) {
      downloadPromises.push(() => getPackages_(maintainerUsername, offset, limit))
    }
  }
  const results = await allSettledQueued(downloadPromises, maxConcurrentDownloads)
  for (const result of results) {
    if (result.status == 'fulfilled') {
      packages.push(...result.value[1])
    } else {
      throw result.reason
    }
  }
  if (publisherUsername) packages = packages.filter(package_ => package_.publisher.username == publisherUsername)
  return packages
}

/**
 * Get the daily download counts of a package (or packages) in the given timeframe. If doing a bulk query then limit it to <=128 packages.
 * @param {string|Array.<string>} packages The package (or packages if an array) to query.
 * @param {string} range A range/timeframe compatible with the registry API.
 * @returns {object|Array.<object>} {packageName, dailyDownloads} (in an array if doing a bulk query)
 */
export async function getLastDownloads(packages, range) {
  if (range == undefined) throw Error('Range must be supplied!')
  let wrapIt
  if (Array.isArray(packages)) {
    if (packages.length > 128) throw Error('Can\'t bulk query more than 128 packages.')
  } else {
    wrapIt = true
    packages = [packages]
  }
  const packageNames = packages.map(package_ => package_.name)
  let queryResult = await getJson('https://api.npmjs.org/downloads/range/'+range+'/'+packageNames.join(','))
  if ('error' in queryResult) throw Error('API error: '+queryResult.error)
  if (wrapIt) {const r = {}; r[queryResult.package] = queryResult; queryResult = r}
  const result = []
  for (const package_ of packages) {
    const packageDetails = queryResult[package_.name]
    if (packageDetails == null) continue // a newly submitted package without stats yet
    const dailyDownloads = packageDetails.downloads.map(o => o.downloads)
    result.push({package: package_, dailyDownloads, endDate: packageDetails.end})
  }
  return result
}

/** Used to keep track of and work with the statistics collected by this library. I will not document its functions very well, instead check the code for usage examples. */
export class PackageStatistics {
  #range; #positionMap; #maintainerUsername; #publisherUsername; #date
  #stats = []

  constructor({maintainerUsername, publisherUsername, range} = {}) {
    this.#range = range
    this.#maintainerUsername = maintainerUsername
    this.#publisherUsername = publisherUsername
  }

  get date() {return this.#date}
  set date(date) {if (!this.#date) this.#date = date}

  /** Add download or score stats (not both at once). */
  add({packageName, dailyDownloads, endDate, score}) {
    if (dailyDownloads) {
      let dailyMin = Number.MAX_SAFE_INTEGER, dailyMax = 0, totalDownloads = 0
      for (const count of dailyDownloads) {
        totalDownloads += count
        if (count > dailyMax) dailyMax = count
        if (count < dailyMin) dailyMin = count
      }
      this.#stats.push({packageName, totalDownloads, dailyDownloads, dailyMin, dailyMax, score})
      this.date = endDate
    } else {
      this.#stats.push({packageName, score})
    }
  }

  /** Sort the stats array. */
  sort(by='downloads') {
    switch (by.toLowerCase()) {
      default: throw Error('Invalid sort by string: '+by)
      case 'downloads':   this.#stats.sort((a, b) => b.totalDownloads - a.totalDownloads); break
      case 'quality':     this.#stats.sort((a, b) => b.score.quality - a.score.quality); break
      case 'popularity':  this.#stats.sort((a, b) => b.score.popularity - a.score.popularity); break
      case 'maintenance': this.#stats.sort((a, b) => b.score.maintenance - a.score.maintenance); break
      case 'final':       this.#stats.sort((a, b) => b.score.final - a.score.final); break
      case 'searchscore': this.#stats.sort((a, b) => b.score.searchScore - a.score.searchScore); break
    }
    this.#createPosMap()
  }

  #createPosMap = function() {
    this.#positionMap = new Map()
    for (let i=0; i<this.#stats.length; i++) {
      this.#positionMap.set(this.#stats[i].packageName, i)
    }
  }

  /** Get the position of the package in the stats array. */
  getPosition(packageName) {
    return this.#positionMap.get(packageName)
  }
  
  /** Get the stats for just this package. */
  getPackageStats(packageName) {
    return this.#stats[this.getPosition(packageName)]
  }

  /** Save stats to disk. */
  saveToDisk(path) {
    const data = {
      range: this.#range,
      date: this.#date,
      maintainerUsername: this.#maintainerUsername,
      publisherUsername: this.#publisherUsername,
      stats: this.#stats
    }
    const json = JSON.stringify(data, null, 2)
    fs.writeFileSync(path, json)
  }

  /** Load stats from disk. */
  loadFromDisk(path) {
    try {
      const json = fs.readFileSync(path, 'utf-8')
      const data = JSON.parse(json)
      this.#range = data.range
      this.#date = data.date
      this.#maintainerUsername = data.maintainerUsername
      this.#publisherUsername = data.publisherUsername
      this.#stats = data.stats
      this.#createPosMap()
      return this
    } catch {}
  }
  
  /** Get a string containing download stats. */
  drawDownloadStats(options, prevDownStats) {
    const chalk = Chalk && !options?.noColors ? Chalk : undefined
    let text = '', range
    switch (this.#range) {
      case 'last-week': range = 'the last '+(chalk ? chalk.yellow('7 days') : '7 days'); break
      case 'last-month': range = 'the last '+(chalk ? chalk.yellow('30 days') : '30 days'); break
      default: range = 'this range '+(chalk ? chalk.yellow(this.#range) : this.#range); break
    }
    if (this.#publisherUsername) {
      text += 'Download count of packages published by '+(chalk ? chalk.yellow(this.#publisherUsername) : this.#publisherUsername)+' '+range+':\n'
    } else {
      text += 'Download count of packages maintained by '+(chalk ? chalk.yellow(this.#maintainerUsername) : this.#maintainerUsername)+' '+range+':\n'
    }
    const climbers = [] // push climbing packages
    for (let i=0; i<this.#stats.length; i++) {
      const pPos = prevDownStats?.date != this.date ? prevDownStats?.getPosition(this.#stats[i].packageName) : undefined
      const positionChange = pPos != undefined ? i - pPos : undefined
      if (positionChange < 0) climbers.push([this.#stats[i].packageName, positionChange])
      text += this.#drawDownloadStat({...this.#stats[i], positionChange}, options) + '\n'
    }
    if (options?.climbersSummary && climbers.length) {
      climbers.sort((a, b) => b - a)
      text += '\nSummary of packages climbing the list:\n'
      for (const [packageName, positionChange] of climbers) {
        text += (chalk ? chalk.green(packageName) : packageName)+' is up: '+Math.abs(positionChange)+'\n'
      }
    }
    if (options?.displayDate) text += '\nLast updated: '+this.date+' UTC (once a day)\n'
    return text.slice(0, -1)
  }

  #drawDownloadStat({packageName, totalDownloads, dailyDownloads, dailyMin, dailyMax, positionChange}, options) {
    let pValue = 0, pChar
    const chalk = Chalk && !options?.noColors ? Chalk : undefined
    function getBarChar(value) {
      const barEven = '⠀⠀⣀⠀⣤⠀⣶⠀⣿'
      const barOddL = '⠀⢀⠀⣠⠀⣴⠀⣾'
      const barOddR = '⠀⢀⠀⣄⠀⣦⠀⣷'
      let remappedValue
      if (dailyMax <= 8) remappedValue = value
      else remappedValue = Math.ceil(8 * (value - dailyMin) / (dailyMax - dailyMin)) // 0 to 8
      if (remappedValue % 2) { // odd
        if (pValue < value) return barOddL[remappedValue]
        if (pValue > value) return barOddR[remappedValue]
        if (pChar) return pChar // if same value; repeat last char
      } else { // even
        return barEven[remappedValue]
      }
    }
    function remap(value) {
      if (value == undefined) return 0
      let remappedValue
      if (dailyMax <= 4) remappedValue = value
      else remappedValue = Math.ceil(4 * (value - dailyMin) / (dailyMax - dailyMin)) // 0 to 4
      return remappedValue
    }
    let line = ''
    function drawGraph() {
      line += '|'
      if (options?.thinBars) {
        for (let i=0; i<dailyDownloads.length; i+=2) {
          line += braille.bars(remap(dailyDownloads[i]), remap(dailyDownloads[i+1]))
        }
      } else {
        for (const count of dailyDownloads) {
          pChar = getBarChar(count)
          line += pChar
          pValue = count
        }
      }
      line += '| '
      if (options?.comparePrevious && typeof positionChange == 'number') {
        const arrow = {
          up: '▲',
          down: '▼'
        }
        if (positionChange > 0) {
          const positionChangeStr = String(positionChange)
          line += chalk ? chalk.red(arrow.down+positionChangeStr) : arrow.down+positionChangeStr
        } else if (positionChange < 0) {
          const positionChangeStr = String(Math.abs(positionChange))
          line += chalk ? chalk.green(arrow.up+positionChangeStr) : arrow.up+positionChangeStr
        }
        if (positionChange != 0) line += ' '
      }
    }
    function drawStats(first) {
      let packageNameString = (chalk ? chalk.green(packageName) : packageName)
      if (terminalLink?.isSupported && !options?.noLinks) {
        packageNameString = terminalLink(packageNameString, 'https://www.npmjs.com/package/'+packageName)
      }
      line += packageNameString + ' '
      if (!options?.noTotal) line += (chalk ? chalk.yellow(totalDownloads) : totalDownloads) + ' '
      if (!options?.noMinMax) line += (chalk ? chalk.blue(`${dailyMin}-${dailyMax}`) : `${dailyMin}-${dailyMax}`)
      line += first ? ': ' : ' '
    }
    if (options?.separateLines) {
      drawStats(true)
      line += '\n'
      drawGraph()
    } else {
      drawGraph()
      drawStats()
    }

    return line
  }

  /** Get a string containing score stats. */
  drawScoreStats(sortedBy, options, prevScoreStats) {
    const chalk = Chalk && !options?.noColors ? Chalk : undefined
    let text = ''
    if (this.#publisherUsername) {
      text += 'Score list of packages published by '+(chalk ? chalk.yellow(this.#publisherUsername) : this.#publisherUsername)
    } else {
      text += 'Score list of packages maintained by '+(chalk ? chalk.yellow(this.#maintainerUsername) : this.#maintainerUsername)
    }
    text += ' sorted by '+(chalk ? chalk.yellow(sortedBy) : sortedBy)+':\n'
    const climbers = [] // push climbing packages
    for (let i=0; i<this.#stats.length; i++) {
      const pPos = prevScoreStats?.date != this.date ? prevScoreStats?.getPosition(this.#stats[i].packageName) : undefined
      const positionChange = pPos != undefined ? i - pPos : undefined
      const pFinal = prevScoreStats?.date != this.date ? prevScoreStats?.getPackageStats(this.#stats[i].packageName).score.final : undefined
      const finalDiff = pFinal != undefined ? this.#stats[i].score.final - pFinal : undefined
      if (positionChange < 0) climbers.push([this.#stats[i].packageName, positionChange])
      text += this.#drawScoreStat({...this.#stats[i], positionChange, finalDiff}, options) + '\n'
    }
    if (options?.climbersSummary && climbers.length) {
      climbers.sort((a, b) => b - a)
      text += '\nSummary of packages climbing the list:\n'
      for (const [packageName, positionChange] of climbers) {
        text += (chalk ? chalk.green(packageName) : packageName)+' is up: '+Math.abs(positionChange)+'\n'
      }
    }
    return text.slice(0, -1)
  }

  #drawScoreStat({packageName, score, positionChange, finalDiff}, options) {
    const chalk = Chalk && !options?.noColors ? Chalk : undefined
    let packageNameString = (chalk ? chalk.green(packageName) : packageName)
    if (terminalLink?.isSupported && !options?.noLinks) {
      packageNameString = terminalLink(packageNameString, 'https://www.npmjs.com/package/'+packageName)
    }
    let text = packageNameString+' '+JSON.stringify(score, (k, v) => {
      if (typeof v == 'number') return wideNumber.format(v)
      return v
    }, 2).replaceAll('"', '')
    if (Math.abs(finalDiff) >= 0.0001) {
      const string = ' (final change: '+wideNumber.format(finalDiff)+')'
      if (finalDiff > 0) text += chalk ? chalk.green(string) : string
      if (finalDiff < 0) text += chalk ? chalk.red(string) : string
    }
    return text
  }
}

async function sha256hex_fromString(message) {
  const {subtle} = (await import('crypto')).webcrypto
  const msgBuffer = new TextEncoder('utf-8').encode(message)
  const hashBuffer = await subtle.digest('SHA-256', msgBuffer)
  return Buffer.from(hashBuffer).toString('hex')
}

/**
 * Get either download or score statistics of all packages where `maintainerUsername` is listed as one of the maintainers (anyone publishing a package is automatically listed as one). Optionally filter out any packages not published by `publisherUsername` (this can be used to return only packages published by this user).
 * @param {object} options 
 * @param {string} options.maintainerUsername
 * @param {string} options.publisherUsername
 * @param {string} [options.range='last-month'] A range/timeframe compatible with the registry API (used for download stats).
 * @param {boolean} [options.withoutDownloads] To only get score stats.
 * @returns {PackageStatistics}
 */
export async function getStatistics({maintainerUsername, publisherUsername, range='last-month', withoutDownloads}) {
  if (!maintainerUsername && publisherUsername) maintainerUsername = publisherUsername
  if (!maintainerUsername) throw Error('Please specify: {maintainerUsername}')
  const packageList = await getPackages(maintainerUsername, publisherUsername)
  const downloadStat = new PackageStatistics({maintainerUsername, publisherUsername, range: withoutDownloads ? undefined : range})
  if (!withoutDownloads) { // with download stats only
    const promiseLaunchers = [] // push functions which when executed returns a promise which downloads the stats
    for (let i=0; i<packageList.length; i+=128) { // do bulk queries (max 128 in one)
      const slicedList = packageList.slice(i, i+128).filter(package_ => {
        if (package_.name.includes('@')) { // scoped packages can't be used in bulk queries
          promiseLaunchers.push(() => getLastDownloads(package_, range)) // put them in single queries
          return false // and filter them out from the bulk query
        }
        return true
      })
      promiseLaunchers.push(() => getLastDownloads(slicedList, range)) // add the bulk query
    }
    const results = await allSettledQueued(promiseLaunchers, maxConcurrentDownloads)
    for (const result of results) {
      if (result.status == 'fulfilled') {
        if (!Array.isArray(result.value)) result.value = [result.value]
        for (const {package: package_, dailyDownloads, endDate} of result.value) {
          downloadStat.add({packageName: package_.name, dailyDownloads, endDate})
        }
      } else {
        throw result.reason
      }
    }
    downloadStat.sort()
  } else { // without download stats
    let scoreHash = ''
    for (const package_ of packageList) {
      scoreHash = await sha256hex_fromString(scoreHash+package_.score.final)
      downloadStat.add({packageName: package_.name, score: {
        final:       package_.score.final,
        ...package_.score.detail,
        searchScore: package_.searchScore
      }})
    }
    downloadStat.date = scoreHash // "date" just needs to be a "unique id" for when stats was updated
  }
  return downloadStat
}

/**
 * For when you want to await lots of promises but not launch them all at once. This function allows you to set a max concurrency to limit how many promises can be launched (and awaited) at the same time. This is a very good way to limit concurrent HTTP requests, etc.
 * @param {Array.<Function>} promiseLaunchers An array with functions that launches and returns a promise.
 * @param {number} [maxConcurrency=1] How many promises that will be allowed to run at the same time.
 * @returns {Promise.<Array.<object>>} An array with objects equal to the result of `Promise.allSettled`.
 * @ignore
 */
async function allSettledQueued(promiseLaunchers, maxConcurrency = 1) {
  const result = [], workerPromises = []

  for (let i=0; i<maxConcurrency; i++) {
    workerPromises.push(new Promise(async resolve => {
      while (promiseLaunchers.length) {
        try {
          result.push({
            status: 'fulfilled',
            value: await promiseLaunchers.pop()() // pop launcher, execute it and await returned promise
          })
        } catch (error) {
          result.push({
            status: 'rejected',
            reason: error
          })
        }
      }
      resolve()
    }))
  }

  await Promise.allSettled(workerPromises)
  return result
}

async function slackWebApi(method, token, payload) {
  const webApiUrl = 'https://slack.com/api/'
  const result = await new Promise((resolve, reject) => {
    const json = JSON.stringify(payload)
    https.request(webApiUrl+method, {method: 'POST', headers: {
      'authorization': 'Bearer '+token,
      'content-type': 'application/json; charset=utf-8'
    }}, response => {
      if (response.statusCode != 200) reject(Error('Response code: '+response.statusCode))
      let json = ''
      response.setEncoding('utf-8')
      response.on('data', chunk => json += chunk)
      response.on('end', () => {
        try {resolve(JSON.parse(json))} 
        catch (error) {reject(error)}
      })
    })
    .on('error', reject)
    .end(json)
  })
  if (result.ok == false) throw Error(result.error)
  if (result.warning) console.warn(result.warning)
  return result
}

function mandatory(parameter, message) {
  throw Error(message+' (missing the `'+parameter+'` parameter)')
}

/**
 * Push provided text to the channel, with customizable username and icon emoji.
 * @param {object} options 
 * @param {string} options.token
 * @param {string} options.channel
 * @param {string} options.text
 * @param {string} [options.username='npm-downloads-counter']
 * @param {string} [options.icon_emoji=':bar_chart:']
 * @returns {Promise.<object>} The result from the Slack Web API.
 */
export function pushToSlack({
  token, channel, text,
  username = 'npm-downloads-counter',
  icon_emoji = ':bar_chart:'
} = {}) {
  token ?? mandatory('token', 'A Slack Web API token is required.')
  channel ?? mandatory('channel', 'Which Slack channel should we push to?')
  text ?? mandatory('text', 'Which text to push to the Slack channel?')
  const parameters = {channel, text, username}
  if (icon_emoji) parameters.icon_emoji = icon_emoji
  return slackWebApi('chat.postMessage', token, parameters)
}
